### PR TITLE
Store domains lists in OrderedSets in networkd

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by yuwata** [comment]
CI fails...
```
589/650 fuzz-network-parser:oss-fuzz-11285:address  FAIL     0.02 s

--- command ---
(snip)
--- stdout ---
/home/runner/systemd/systemd/test/fuzz/fuzz-network-parser/oss-fuzz-11285... 
--- stderr ---
Failed to read $container of PID 1, ignoring: Permission denied
Found container virtualization none.
Assertion 's' failed at ../../../../src/basic/ordered-set.c:21, function ordered_set_put_strdup(). Aborting.
-------
```

File: src/network/networkd-network.c [hunk 1]
```diff
@@ -1219,15 +1218,11 @@ int config_parse_radv_search_domains(
...
-                        r = strv_push(&n->router_search_domains, idna);
-                        if (r >= 0)
-                                idna = NULL;
-                } else {
-                        r = strv_push(&n->router_search_domains, w);
-                        if (r >= 0)
-                                w = NULL;
-                }
+
+                r = ordered_set_put_strdup(n->router_search_domains,
```
Comment (yuwata) [inline src/network/networkd-network.c]: `ensure_allocated()` may be necessary before calling this.

File: src/network/networkd-network.c [hunk 2]
```diff
@@ -1171,32 +1199,30 @@ int config_parse_radv_search_domains(
...
+                                           TAKE_PTR(idna) ?: TAKE_PTR(w));
```
Comment (yuwata) [inline src/network/networkd-network.c]: I think we should not do that. `ordered_set_put_strdup()` may fail.

File: src/test/test-ordered-set.c [hunk 3]
```diff
@@ -61,12 +63,47 @@ static void test_set_put(void) {
         assert_se(ordered_set_put(m, (void*) "333") == 1);
         assert_se(ordered_set_put(m, (void*) "333") == 0);
         assert_se(ordered_set_put(m, (void*) "22") == 0);
+
+        _cleanup_free_ char **t = ordered_set_get_strv(m);
```
Comment (poettering) [inline src/test/test-ordered-set.c]: not that it matters much, but there should be an `assert_se(t)` here, i figure

File: src/test/test-ordered-set.c [hunk 4]
```diff
@@ -61,12 +63,47 @@ static void test_set_put(void) {
...
+        assert_se(t = strdup("333"));
+        assert_se(ordered_set_put(m, t) == 1);
+
+        assert_se(ordered_set_put(q, (void*) "11") == 1);
+        assert_se(ordered_set_put(q, (void*) "22") == 1);
+        assert_se(ordered_set_put(q, (void*) "33") == 1);
+
+        assert_se(ordered_set_put_string_set(m, q) == 2);
+
+        _cleanup_free_ char **final = ordered_set_get_strv(m);
```
Comment (poettering) [inline src/test/test-ordered-set.c]: as above

File: src/test/test-set.c [hunk 5]
```diff
@@ -77,6 +78,12 @@ static void test_set_put(void) {
         assert_se(set_put(m, (void*) "333") == 1);
         assert_se(set_put(m, (void*) "333") == 0);
         assert_se(set_put(m, (void*) "22") == 0);
+
+        _cleanup_free_ char **t = set_get_strv(m);
```
Comment (poettering) [inline src/test/test-set.c:82]: as above

**Comment by poettering** [comment]
travis CI doesn't like this. and looks like it is right.

**Comment by keszybz** [comment]
Updated with the requested asserts added. Travis was just complaining about an unused variable. Let's see if everything passes now.

**Comment by keszybz** [comment]
It is failing with a timeout. The timeout is 30s. Dunno, maybe the right solution is simply to shorten the test. oss-fuzz will test us if the timeout on their side is fixed.

**Pull request merged**